File: LeetCode/Easy/0021. Merge Two Sorted Lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */



struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2){
    struct ListNode *lst = NULL;     
    
    while(list1 || list2) {
        struct ListNode *newNode = malloc(sizeof(struct ListNode));
        newNode->next = NULL;
        
        if(!list1) {
            newNode->val = list2->val;
            list2 = list2->next;
        } else if(!list2) {
            newNode->val = list1->val;
            list1 = list1->next;
        } else if(list1->val <= list2->val) {
            newNode->val = list1->val;
            list1 = list1->next;
        } else {
            newNode->val = list2->val;
            list2 = list2->next;
        }
        
        struct ListNode *curr = lst;
        if(curr == NULL) {
            lst = newNode;
        } else {
            while(curr->next) {
                curr = curr->next;
            }
            curr->next = newNode;
        }
    }
    
    return lst;
}

File: LeetCode/Easy/0021. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """

        curr_node, start_node = None, None
        while list1 or list2:
            if list1 == None:
                new_node = list2
                list2 = new_node.next
            elif list2 == None:
                new_node = list1
                list1 = new_node.next
            elif list1.val <= list2.val:
                new_node = list1
                list1 = new_node.next
            else:
                new_node = list2
                list2 = new_node.next
            
            if curr_node == None:
                curr_node = new_node
                start_node = curr_node
            else:
                curr_node.next = new_node
                curr_node = new_node
        
        return start_node

        
